sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala

```diff
@@ -723,4 +723,18 @@ class DataFrameAggregateSuite extends QueryTest with SharedSQLContext {
       "grouping expressions: [current_date(None)], value: [key: int, value: string], " +
         "type: GroupBy]"))
   }
+
+  test("SPARK-26021: Double and Float 0.0/-0.0 should be equal when grouping") {
+    val colName = "i"
+    val doubles = Seq(0.0d, 0.0d, -0.0d).toDF(colName).groupBy(colName).count().collect()
```

Comment:
Without this change, `doubles` already has expected results:

```scala
Seq(0.0d, 0.0d, -0.0d).toDF(colName).groupBy(colName).count().show()

+---+-----+
|  i|count|
+---+-----+
|0.0|    3|
+---+-----+
```

Do you know why?

Reply:
Actually yes, if codegen is enabled a generated FastHashMap is used for the partial grouping before the shuffle. This map doesn't separate 0.0 and -0.0. In addition there are 2 threads for 2 partitions in the unit test. In the doubles Seq order each partition is grouped in 0.0 and so after the shuffle they are being merged.
In the floats case the order of the elements in the Seq is different so in the first grouping we get 1 partition on 0.0 and the other on -0.0 and so after the shuffle they are being treated as different groups (before the fix).
Now I remember this is the reason I originally disabled codegen in the test.
I think I'll just reorder the doubles Seq as well so it manifests the bug.

Reply:
Yeah, it is better to reorder the doubles.

Reply:
Example:
```
Seq(0.0d, 0.0d, -0.0d).toDF(colName).groupBy(colName).count().show()
+---+-----+
|  i|count|
+---+-----+
|0.0|    3|
+---+-----+
Seq(0.0d, -0.0d, 0.0d).toDF(colName).groupBy(colName).count().show()
+----+-----+
|   i|count|
+----+-----+
| 0.0|    1|
|-0.0|    2|
+----+-----+
Seq(-0.0d, -0.0d, 0.0d).toDF(colName).groupBy(colName).count().show()
+----+-----+
|   i|count|
+----+-----+
|-0.0|    3|
+----+-----+
```